File: Strand/Data/HydrationStore.swift
import Foundation
import WhoopStore
import StrandAnalytics

// MARK: - Hydration tracker (MVP) — opt-in, local-only water logging
//
// The user logs water with three quick taps (Sip 30 ml / Cup 237 ml / Bottle 500 ml). The day TOTAL is
// banked in the generic metric-series tall table under a dedicated source/key — the SAME `metricSeries`
// table + `upsertMetricSeries` path every other generic daily series uses (no schema change). Because the
// table holds one row per (deviceId, day, key), a tap reads the day's running total and re-upserts
// total + amount, so the stored value IS "the sum of today's hydration logged for this local day".
//
// This is the BYTE-PARITY twin of the Android `com.noop.analytics.HydrationStore`: identical source id
// ("hydration"), identical key ("hydration"), identical additive-accumulation logic, identical 7-day
// history projection (one row per local calendar day, 0 for empty days, oldest first). Per-tap timestamps
// are intentionally NOT persisted on either platform — the day total is the source of truth and the MVP
// detail shows the honest day figure. Everything stays on-device; nothing is synced.

enum HydrationStore {
    /// Source/device id the hydration total is written under — its own local-only source so it is never
    /// confused with strap-imported or computed metrics. MUST match the Android `SOURCE_ID`.
    static let sourceId = "hydration"

    /// metricSeries key for the daily total (ml). MUST match the Android `KEY`.
    static let key = "hydration"

    /// Settings opt-in key (default OFF). The dashboard card + detail are hidden while this is false.
    /// MUST match the Android `NoopPrefs.KEY_HYDRATION_TRACKING` so the toggle reads the same on both.
    static let enabledKey = "noop.hydrationTracking"

    /// UserDefaults prefix for the per-day entry list (#798). One JSON array per local day, keyed
    /// `noop.hydrationEntries.<yyyy-MM-dd>`. Local-only, on-device, never synced - the same privacy posture
    /// as the day total. The day total in `metricSeries` stays the canonical figure the rest of the app
    /// reads (Today card, ring, 7-day history); the entry list is the editable detail behind it, kept in
    /// sync so deleting/editing an entry re-derives and re-banks the total.
    static let entriesKeyPrefix = "noop.hydrationEntries."

    /// AppStorage key for the user's custom container size (ml) (#798). Default `cupML` until set.
    static let customSizeKey = "noop.hydrationCustomSizeML"

    /// metricSeries key for water IMPORTED from the platform health store — Apple Health on iOS,
    /// Health Connect on Android (#949). MUST match the Android `KEY_IMPORTED`.
    ///
    /// Kept in its own row rather than folded into `key` so the two can never be confused, because they
    /// behave differently on write: `key` ACCUMULATES (each tap adds), while this one is REPLACED with
    /// the platform's recomputed day sum on every sync. That replacement is what makes re-importing
    /// idempotent without tracking a single sample id — HealthKit's `HKStatisticsCollectionQuery` and
    /// Health Connect's day bucket both hand back the whole day's total, so writing it wholesale means a
    /// second sync of the same day stores the same number, and a drink deleted in the source app makes
    /// the figure go DOWN on the next sync instead of being stranded forever.
    ///
    /// The user's own taps are never touched by a sync, and imported water is never editable here — it
    /// is owned by the app that logged it.
    static let importedKey = "hydrationImported"

    static func entriesKey(forDay dayKey: String) -> String { entriesKeyPrefix + dayKey }
}

// MARK: - Per-entry model (#798) - individual logged drinks for edit/delete

/// One logged drink: a stable id, the amount (ml) and the wall-clock time it was logged. Local-only;
/// persisted as a JSON array per local day. `Codable`/`Equatable` so the list round-trips through
/// UserDefaults and is unit-testable.
struct HydrationEntry: Identifiable, Equatable, Codable {
    let id: UUID
    var amountMl: Int
    var loggedAt: Date

    init(id: UUID = UUID(), amountMl: Int, loggedAt: Date = Date()) {
        self.id = id
        self.amountMl = amountMl
        self.loggedAt = loggedAt
    }
}

/// Pure list operations over the per-day entries (#798). Kept free of persistence/UI so the add / delete /
/// edit / total math is unit-testable in isolation. The day total is ALWAYS the sum of the (clamped to ≥ 0)
/// entry amounts, so deleting or editing an entry can only ever produce a non-negative, self-consistent total.
enum HydrationEntries {
    /// Append a new entry. A non-positive amount is rejected (returns the list unchanged) so a stray 0/negative
    /// can never enter the list, matching `logHydration`'s no-op-on-non-positive contract.
    static func adding(_ entries: [HydrationEntry], amountMl: Int, at date: Date = Date()) -> [HydrationEntry] {
        guard amountMl > 0 else { return entries }
        return entries + [HydrationEntry(amountMl: amountMl, loggedAt: date)]
    }

    /// Remove the entry with `id` (a no-op if absent).
    static func removing(_ entries: [HydrationEntry], id: UUID) -> [HydrationEntry] {
        entries.filter { $0.id != id }
    }

    /// Set an existing entry's amount. A non-positive amount removes the entry (an edit to 0 is a delete),
    /// keeping the list free of zero rows. Unknown ids are ignored.
    static func updating(_ entries: [HydrationEntry], id: UUID, amountMl: Int) -> [HydrationEntry] {
        guard amountMl > 0 else { return removing(entries, id: id) }
        return entries.map { $0.id == id ? HydrationEntry(id: $0.id, amountMl: amountMl, loggedAt: $0.loggedAt) : $0 }
    }

    /// The day total (ml) = sum of the entry amounts, each clamped ≥ 0. Always non-negative.
    static func total(_ entries: [HydrationEntry]) -> Double {
        entries.reduce(0) { $0 + Double(max(0, $1.amountMl)) }
    }
}

// MARK: - Logging + read seam (Repository extension)

extension Repository {

    /// The total fluid (ml) for a local day (yyyy-MM-dd) as the user should SEE it: what they logged by
    /// hand plus whatever was imported from Apple Health / Health Connect (#949). 0 when neither exists.
    /// Mirrors Android `HydrationStore.total`.
    ///
    /// Everything that DISPLAYS a day figure (Today card, ring, detail) wants this. Everything that
    /// WRITES a hand-logged drink must use `hydrationManualTotal` instead — see there.
    func hydrationTotal(day: String) async -> Double {
        await hydrationManualTotal(day: day) + hydrationImportedTotal(day: day)
    }

    /// Only what the user logged by hand — the row `logHydration` accumulates into.
    ///
    /// The write path MUST read this and not `hydrationTotal`: a quick-add stores `current + amount`, so
    /// reading the combined figure would fold every imported millilitre into the manual row, and the next
    /// sync would then add the imported water on top of the copy it had just made. One tap after an
    /// import would silently double the day, and it would compound on every tap after that.
    func hydrationManualTotal(day: String) async -> Double {
        await metricSeriesValue(key: HydrationStore.key, day: day)
    }

    /// Only what came from the platform health store (#949). Replaced wholesale by each sync.
    func hydrationImportedTotal(day: String) async -> Double {
        await metricSeriesValue(key: HydrationStore.importedKey, day: day)
    }

    /// One day's value for a hydration-source series, or 0 when the row is absent.
    private func metricSeriesValue(key: String, day: String) async -> Double {
        guard let store = await storeHandle() else { return 0 }
        let pts = (try? await store.metricSeries(deviceId: HydrationStore.sourceId,
                                                 key: key, from: day, to: day)) ?? []
        return pts.first?.value ?? 0
    }

    /// Replace the imported-water total for each given local day with the platform's recomputed day sum
    /// (#949). Called by the Apple Health sync; the Android twin is `HydrationStore.setImported`.
    ///
    /// REPLACES rather than adds, which is the whole reason this is idempotent — see `importedKey`. Days
    /// the health store has no water for resolve to 0, so water deleted in the source app disappears here
    /// too instead of lingering as a stale row.
    ///
    /// Only days whose value actually MOVED are written. The caller hands over its whole window (31 days
    /// on iOS) on every sync, and `sync` is driven by six hourly `HKObserverQuery` wakes — so writing it
    /// wholesale meant ~31 upserts plus a `hydrationSeq` bump, repeatedly through the day, to store
    /// numbers that were already there. Thirty of those days are historical and immutable; the usual
    /// honest answer is "nothing changed". One ranged read replaces the write burst.
    ///
    /// A day with NO row and a value of 0 is skipped rather than written, because an absent row already
    /// reads as 0 (`hydrationImportedTotal`). That is what keeps a user with no water in Health from
    /// accumulating a wall of zero rows. A day that HAS a row and is now 0 is still written — that is the
    /// deletion propagating, and it must not be mistaken for the no-op case.
    func setImportedHydration(_ mlByDay: [String: Double]) async {
        guard !mlByDay.isEmpty, let store = await storeHandle() else { return }
        let days = mlByDay.keys.sorted()
        guard let from = days.first, let to = days.last else { return }

        // No `?? []` here: a FAILED read must not look like "there are no rows". If it did, a day whose
        // water was just deleted would resolve to the skip-a-zero case below and its stale non-zero row
        // would survive. Bail instead and let the next sync do it properly — the health store is the
        // source of truth, so nothing is lost by waiting.
        guard let existing = try? await store.metricSeries(deviceId: HydrationStore.sourceId,
                                                           key: HydrationStore.importedKey,
                                                           from: from, to: to) else { return }
        var stored: [String: Double] = [:]
        for p in existing { stored[p.day] = p.value }

        // Return type spelled out rather than inferred: this file is app-target Swift that no CI
        // compiles, so a multi-statement closure leaning on inference is a needless place to be wrong.
        let changed: [MetricPoint] = mlByDay.compactMap { (day, ml) -> MetricPoint? in
            let value = max(0, ml)
            guard let current = stored[day] else {
                // No row yet: only worth creating one if there is actually something to record.
                return value == 0 ? nil : MetricPoint(day: day, key: HydrationStore.importedKey, value: value)
            }
            // Epsilon rather than ==: the same samples re-summed give the same Double, but a value that
            // ever drifted by a float hair would otherwise re-write every sync forever and quietly undo
            // the point of this. Well below a millilitre, so no real change is swallowed.
            guard abs(current - value) > 1e-9 else { return nil }
            return MetricPoint(day: day, key: HydrationStore.importedKey, value: value)
        }
        guard !changed.isEmpty else { return }

        _ = try? await store.upsertMetricSeries(changed, deviceId: HydrationStore.sourceId)
        // Only on a real change — this bump re-reads the Today card, and firing it on every sync made the
        // card redo its hydration read for nothing. Same reason as logHydration: hydration writes never
        // bump refreshSeq, so the card has no other signal.
        noteHydrationChanged()
    }

    /// Log `amountMl` of fluid for `day` (defaults to today's local day). Reads the day's current total
    /// and upserts total + amount, so repeated taps accumulate. A non-positive amount is a no-op. Returns
    /// the new day total (ml). Additive by design — each tap is a quick-add, like the WHOOP buttons.
    /// Mirrors Android `HydrationStore.log`.
    @discardableResult
    func logHydration(amountMl: Int, day: String? = nil) async -> Double {
        let dayKey = day ?? Repository.localDayKey(Date())
        guard amountMl > 0, let store = await storeHandle() else { return await hydrationTotal(day: dayKey) }
        // MANUAL total, never the combined one (#949) — see `hydrationManualTotal`.
        let current = await hydrationManualTotal(day: dayKey)
        let next = current + Double(amountMl)
        _ = try? await store.upsertMetricSeries(
            [MetricPoint(day: dayKey, key: HydrationStore.key, value: next)],
            deviceId: HydrationStore.sourceId)
        // #798 - also record the per-entry row so the detail can show, edit and delete this exact drink.
        let entries = HydrationEntries.adding(Self.hydrationEntries(day: dayKey), amountMl: amountMl)
        Self.writeHydrationEntries(entries, day: dayKey)
        // #989: hydration writes never bump refreshSeq, so tell the Today card directly.
        noteHydrationChanged()
        // `next` is the manual row; callers of this return value display it, so hand back the combined
        // figure (#949). Every in-tree caller discards it and re-reads, but a caller that trusted it
        // would otherwise show a total that drops the imported water.
        return next + (await hydrationImportedTotal(day: dayKey))
    }

    // MARK: - Per-entry edit/delete (#798)

    /// Today (or `day`)'s individual logged drinks, oldest first, as persisted in UserDefaults. Empty when
    /// nothing has been logged that day. Local-only; never synced.
    func hydrationEntries(day: String? = nil) -> [HydrationEntry] {
        Self.hydrationEntries(day: day ?? Repository.localDayKey(Date()))
    }

    /// Delete one logged entry by id, then re-derive the day total from the surviving entries and re-bank it
    /// into `metricSeries` so the ring, Today card and 7-day history all reflect the deletion. Returns the
    /// new day total (ml).
    @discardableResult
    func deleteHydrationEntry(id: UUID, day: String? = nil) async -> Double {
        let dayKey = day ?? Repository.localDayKey(Date())
        let next = HydrationEntries.removing(Self.hydrationEntries(day: dayKey), id: id)
        Self.writeHydrationEntries(next, day: dayKey)
        return await rebankHydrationTotal(entries: next, day: dayKey)
    }

    /// Set an existing entry's amount (a non-positive amount deletes it), then re-derive + re-bank the day
    /// total. Returns the new day total (ml). Backs the "edit a logged drink / set a custom size" flow.
    @discardableResult
    func updateHydrationEntry(id: UUID, amountMl: Int, day: String? = nil) async -> Double {
        let dayKey = day ?? Repository.localDayKey(Date())
        let next = HydrationEntries.updating(Self.hydrationEntries(day: dayKey), id: id, amountMl: amountMl)
        Self.writeHydrationEntries(next, day: dayKey)
        return await rebankHydrationTotal(entries: next, day: dayKey)
    }

    /// Re-derive the day total from `entries` and upsert it into `metricSeries` (the canonical total). The
    /// per-entry list is the source of truth for an edited/deleted day; this keeps the rest of the app in sync.
    @discardableResult
    private func rebankHydrationTotal(entries: [HydrationEntry], day dayKey: String) async -> Double {
        let total = HydrationEntries.total(entries)
        if let store = await storeHandle() {
            _ = try? await store.upsertMetricSeries(
                [MetricPoint(day: dayKey, key: HydrationStore.key, value: total)],
                deviceId: HydrationStore.sourceId)
        }
        // #989: edits/deletes funnel through here; tell the Today card directly (see logHydration).
        noteHydrationChanged()
        // The entry list only ever describes hand-logged drinks, so re-deriving from it must not be
        // allowed to erase the imported row — write the manual figure, return the combined one (#949).
        return total + (await hydrationImportedTotal(day: dayKey))
    }

    // MARK: - Entry persistence (UserDefaults JSON, one array per local day)

    fileprivate static func hydrationEntries(day dayKey: String) -> [HydrationEntry] {
        guard let data = UserDefaults.standard.data(forKey: HydrationStore.entriesKey(forDay: dayKey)),
              let decoded = try? JSONDecoder().decode([HydrationEntry].self, from: data) else { return [] }
        return decoded.sorted { $0.loggedAt < $1.loggedAt }
    }

    fileprivate static func writeHydrationEntries(_ entries: [HydrationEntry], day dayKey: String) {
        let key = HydrationStore.entriesKey(forDay: dayKey)
        if entries.isEmpty {
            UserDefaults.standard.removeObject(forKey: key)
        } else if let data = try? JSONEncoder().encode(entries) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }

    /// The last `days` local-day totals up to and including today, OLDEST first, as (day, ml) pairs — one
    /// entry per calendar day with 0 for days that have no log. Backs the 7-day mini bar history. `days`
    /// is clamped ≥ 1. Mirrors Android `HydrationStore.history` (a single ranged read projected onto the
    /// full day grid so empty days read as 0 rather than vanishing).
    func hydrationHistory(days: Int = 7, now: Date = Date()) async -> [(day: String, value: Double)] {
        let n = max(1, days)
        let from = now.addingTimeInterval(-Double(n - 1) * 86_400)
        let fromKey = Repository.localDayKey(from)
        let toKey = Repository.localDayKey(now)
        var byDay: [String: Double] = [:]
        if let store = await storeHandle() {
            // Both rows, summed per day (#949) — the bars have to agree with the Today card and the
            // ring, and those read the combined `hydrationTotal`. Reading only the manual key here is
            // what would make an imported day render short.
            for key in [HydrationStore.key, HydrationStore.importedKey] {
                let pts = (try? await store.metricSeries(deviceId: HydrationStore.sourceId,
                                                         key: key,
                                                         from: fromKey, to: toKey)) ?? []
                for p in pts { byDay[p.day, default: 0] += p.value }
            }
        }
        return (0..<n).map { i in
            let key = Repository.localDayKey(now.addingTimeInterval(-Double(n - 1 - i) * 86_400))
            return (key, byDay[key] ?? 0)
        }
    }

    /// Today's hydration goal (ml) from the profile sex + today's Effort score. Pure math in
    /// `HydrationGoal`; this just feeds it the live inputs (today's `strain` is NOOP's 0–100 Effort).
    func hydrationGoalML(profileSex: String) -> Int {
        HydrationGoal.dailyGoalML(sex: profileSex, effort: today?.strain)
    }
}

File: android/app/src/main/java/com/noop/analytics/HydrationStore.kt
package com.noop.analytics

import com.noop.data.MetricSeriesRow
import com.noop.data.WhoopRepository
import java.util.TimeZone

/**
 * HydrationStore — the logging + read seam for the Hydration tracker (MVP, opt-in, local-only).
 *
 * Kotlin twin of the Swift hydration store calls. The day total is banked in the generic metric-series
 * store under the [KEY] series, keyed by the device's LOCAL calendar day — the SAME `metricSeries`
 * table + `WhoopRepository.upsertMetricSeries` path every other generic daily series uses (no schema
 * change). Because that table holds one row per (deviceId, day, key), a tap reads the day's running
 * total and re-upserts total + amount, so the stored value IS "the sum of today's hydration logged for
 * this local day". Everything stays on-device; nothing is synced.
 *
 * `ts` (a wall-clock unix second) selects which local day a log lands on; the goal itself comes from the
 * pure [HydrationGoal] engine, never from here.
 */
object HydrationStore {

    /**
     * #989 (Kotlin twin of Repository.hydrationSeq): bumped on every mutation ([log] / [set]; [remove]
     * routes through [set]). Hydration writes never touch the flows Today already collects (`days` only
     * changes on a data refresh), so the dashboard card sat stale until an unrelated sync. Today keys its
     * hydration re-read on this too.
     */
    val mutationSeq = kotlinx.coroutines.flow.MutableStateFlow(0)

    /** The generic metric-series key the day total is banked under (shared id; keep == the Swift key). */
    const val KEY: String = "hydration"

    /** The source/device id the hydration total is written under — its own local-only source so it is
     *  never confused with strap-imported or computed metrics. Matches the Swift source id. */
    const val SOURCE_ID: String = "hydration"

    /**
     * The series key for water IMPORTED from the platform health store — Health Connect here, Apple
     * Health on iOS (#949). Keep == the Swift `HydrationStore.importedKey`.
     *
     * Its own row rather than folded into [KEY], because the two behave differently on write: [KEY]
     * ACCUMULATES (each tap adds to it), while this one is REPLACED with the health store's recomputed
     * day sum on every import. That replacement is what makes re-importing idempotent without tracking
     * individual sample ids — Health Connect hands back every record in the window, so summing them and
     * storing the result wholesale means a second import of the same day stores the same number, and a
     * drink deleted in the source app makes the figure go DOWN next time rather than being stranded.
     *
     * Hand-logged water is never touched by an import, and imported water is not editable here — it is
     * owned by the app that logged it.
     */
    const val KEY_IMPORTED: String = "hydrationImported"

    /** Seconds EAST of UTC for the device's current zone — the offset [AnalyticsEngine.dayString] needs
     *  to bucket a timestamp on the LOCAL calendar day (matches the dashboard's local "today" read). */
    private fun localOffsetSec(atMillis: Long = System.currentTimeMillis()): Long =
        (TimeZone.getDefault().getOffset(atMillis) / 1000).toLong()

    /** The LOCAL yyyy-MM-dd day key for a unix-seconds [ts] (defaults to now). */
    fun dayKey(ts: Long = System.currentTimeMillis() / 1000L): String =
        AnalyticsEngine.dayString(ts, localOffsetSec(ts * 1000L))

    /**
     * Log [amountMl] of fluid for the local day containing [ts] (defaults to now). Reads the day's
     * current total and upserts total + amount under [SOURCE_ID]/[KEY], so repeated taps accumulate.
     * A non-positive amount is a no-op. Returns the new day total (ml). Idempotency is by design absent —
     * each tap is an additive log, matching the WHOOP-style quick-add buttons.
     */
    suspend fun log(repo: WhoopRepository, amountMl: Int, ts: Long = System.currentTimeMillis() / 1000L): Double {
        if (amountMl <= 0) return total(repo, ts)
        val day = dayKey(ts)
        // The MANUAL row, never the combined figure (#949): a quick-add stores `current + amount`, so
        // reading the combined total would copy every imported millilitre into the hand-logged row, and
        // the next import would then add the imported water on top of that copy. One tap after an import
        // would silently double the day, compounding on every tap after it.
        val current = manualTotal(repo, ts)
        val next = current + amountMl
        repo.upsertMetricSeries(listOf(MetricSeriesRow(SOURCE_ID, day, KEY, next)))
        mutationSeq.value += 1   // #989: tell Today's card directly (see mutationSeq)
        return next + importedTotal(repo, ts)
    }

    /**
     * Pure clamp behind [set]: a day total can never be negative. Factored out so the correction math
     * (#798) is unit-testable without a Room/repo stand-in. Returns [totalMl] floored at 0.0.
     */
    fun clampedTotal(totalMl: Double): Double = totalMl.coerceAtLeast(0.0)

    /**
     * Pure result of removing [amountMl] from a [currentTotalMl] (#798): a non-positive amount is a no-op
     * (the current total, still clamped at 0), otherwise the difference floored at 0 so an over-subtraction
     * lands on an empty day rather than a negative total. The testable core of [remove].
     */
    fun afterRemoving(currentTotalMl: Double, amountMl: Int): Double =
        if (amountMl <= 0) clampedTotal(currentTotalMl) else clampedTotal(currentTotalMl - amountMl)

    /**
     * Set the day total directly to [totalMl] for the local day containing [ts], clamped at 0 (a negative
     * target lands on 0, never a negative total). The correction seam behind the detail screen's
     * delete/undo affordances (#798): because the schema banks ONE additive total per (source, day, key)
     * row, an entry isn't separately addressable - removing or editing a log is expressed as adjusting the
     * day total. Returns the new stored total (ml). Mirrors the iOS `setHydration`.
     */
    suspend fun set(repo: WhoopRepository, totalMl: Double, ts: Long = System.currentTimeMillis() / 1000L): Double {
        val day = dayKey(ts)
        val next = clampedTotal(totalMl)
        // Sets the HAND-LOGGED row only (#949). Imported water is not the user's to correct from here —
        // it belongs to the app that logged it, and an import would overwrite the edit on the next run.
        repo.upsertMetricSeries(listOf(MetricSeriesRow(SOURCE_ID, day, KEY, next)))
        mutationSeq.value += 1   // #989: edits/deletes route through here too
        return next + importedTotal(repo, ts)
    }

    /**
     * Remove [amountMl] from the local day's running total (the undo / delete-a-log path for the detail
     * screen, #798). Subtracts the amount and clamps at 0 so the total never goes negative; a non-positive
     * amount is a no-op. Returns the new day total (ml). Built on [set] + [afterRemoving] so the correction
     * math is shared + tested. Mirrors the iOS `removeHydration`.
     */
    suspend fun remove(repo: WhoopRepository, amountMl: Int, ts: Long = System.currentTimeMillis() / 1000L): Double {
        if (amountMl <= 0) return total(repo, ts)
        // Subtract from the MANUAL row (#949). Against the combined figure this would be badly wrong:
        // with 200 ml logged by hand and 500 ml imported, removing 100 would compute 700-100=600 and
        // store THAT as the hand-logged total — inflating the day to 1100 instead of reducing it to 600.
        return set(repo, afterRemoving(manualTotal(repo, ts), amountMl), ts)
    }

    /**
     * The total fluid (ml) for the local day containing [ts] as the user should SEE it: hand-logged plus
     * whatever was imported from the health store (#949). 0.0 when neither exists.
     *
     * Everything that DISPLAYS a day figure wants this. Everything that WRITES a hand-logged amount must
     * use [manualTotal] — see [log] and [remove].
     */
    suspend fun total(repo: WhoopRepository, ts: Long = System.currentTimeMillis() / 1000L): Double =
        manualTotal(repo, ts) + importedTotal(repo, ts)

    /** Only what the user logged by hand — the row [log] accumulates into and [set] overwrites. */
    suspend fun manualTotal(repo: WhoopRepository, ts: Long = System.currentTimeMillis() / 1000L): Double =
        dayValue(repo, KEY, dayKey(ts))

    /** Only what came from the health store (#949). Replaced wholesale by each import. */
    suspend fun importedTotal(repo: WhoopRepository, ts: Long = System.currentTimeMillis() / 1000L): Double =
        dayValue(repo, KEY_IMPORTED, dayKey(ts))

    private suspend fun dayValue(repo: WhoopRepository, key: String, day: String): Double =
        repo.metricSeries(SOURCE_ID, key, day, day).firstOrNull()?.value ?: 0.0

    /**
     * Pure: sum metric rows into ml-per-day (#949). Rows for the same day ADD, which is the whole point —
     * the hand-logged and imported series are two rows on the same day, and a plain `associate { }` would
     * keep only the last one and silently drop the other from every bar in the history chart.
     */
    fun sumByDay(rows: List<MetricSeriesRow>): Map<String, Double> {
        val out = HashMap<String, Double>()
        for (r in rows) out[r.day] = (out[r.day] ?: 0.0) + r.value
        return out
    }

    /**
     * Pure: what an import should WRITE for a window (#949), given the days it covers and the ml it
     * actually found. Every day in [windowDays] gets a value — days with no water resolve to 0.0 rather
     * than being omitted, which is what makes a drink deleted in the source app disappear here instead of
     * leaving the old figure stranded forever.
     *
     * Days found outside the window are dropped: writing them would be a partial update of a day this
     * import never fully examined, so the figure could not be trusted as a replacement.
     */
    fun importWindow(windowDays: List<String>, found: Map<String, Double>): Map<String, Double> =
        windowDays.associateWith { clampedTotal(found[it] ?: 0.0) }

    /**
     * Replace the imported-water total for each (dayKey → ml) pair with the health store's recomputed day
     * sum (#949). Called by the Health Connect import; the iOS twin is `Repository.setImportedHydration`.
     *
     * REPLACES rather than adds — that is what makes re-importing idempotent (see [KEY_IMPORTED]). The
     * caller is expected to pass 0.0 for days it found no water on, so a drink deleted in the source app
     * disappears here too instead of lingering as a stale row.
     */
    suspend fun setImported(repo: WhoopRepository, mlByDay: Map<String, Double>) {
        if (mlByDay.isEmpty()) return
        repo.upsertMetricSeries(
            mlByDay.map { (day, ml) -> MetricSeriesRow(SOURCE_ID, day, KEY_IMPORTED, clampedTotal(ml)) },
        )
        mutationSeq.value += 1   // #989: Today's card reads hydration off this
    }

    /**
     * The last [days] local-day totals up to and including today, OLDEST first, as (dayKey, ml) pairs —
     * one entry per calendar day with 0.0 for days that have no log. Backs the detail screen's 7-day
     * mini bar history. [days] is clamped ≥ 1.
     */
    suspend fun history(
        repo: WhoopRepository,
        days: Int = 7,
        nowSec: Long = System.currentTimeMillis() / 1000L,
    ): List<Pair<String, Double>> {
        val n = days.coerceAtLeast(1)
        val from = nowSec - (n - 1).toLong() * 86_400L
        val fromKey = dayKey(from)
        val toKey = dayKey(nowSec)
        // Both rows, summed per day via [sumByDay] (#949) — the bars have to agree with the Today card
        // and the ring, and those read the combined [total]. Reading only [KEY] renders an imported day
        // short. Projected onto the full day grid below so empty days read as 0 rather than vanishing.
        val byDay = sumByDay(
            repo.metricSeries(SOURCE_ID, KEY, fromKey, toKey) +
                repo.metricSeries(SOURCE_ID, KEY_IMPORTED, fromKey, toKey),
        )
        return (0 until n).map { i ->
            val key = dayKey(nowSec - (n - 1 - i).toLong() * 86_400L)
            key to (byDay[key] ?: 0.0)
        }
    }
}
